File: 5.prefix-sum/동규/2.product-of-array-except-self.js
// https://leetcode.com/problems/product-of-array-except-self/

var productExceptSelf = function (nums) {
  let result = [];
  let lp = 1,
    rp = 1;
  for (let i = 0; i < nums.length; i++) {
    result.push(lp);
    lp *= nums[i];
  }
  for (let i = nums.length - 1; i >= 0; i--) {
    result[i] *= rp;
    rp *= nums[i];
  }
  return result;
};

File: 5.prefix-sum/연경/2.product-of-array-except-self.py
## https://leetcode.com/problems/product-of-array-except-self/
from typing import List
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:

        answer = [1 for _ in range(len(nums))]
        n = 1
        for i in range(len(nums)):
            answer[i] = n
            n *= nums[i]

        n = 1
        for i in range(len(nums)-1, -1, -1):
            answer[i] *= n
            n *= nums[i]

        return answer

if __name__ == "__main__":
    nums = [-1,1,0,-3,3]
    print(Solution().productExceptSelf(nums))